crates/isograph_schema/src/validate_client_field.rs

```diff
@@ -619,43 +632,352 @@ fn get_missing_arguments_and_validate_argument_types<'a>(
+                }
+            }
+        }
+    }
+}
+
+fn id_annotation_to_typename_annotation(
+    type_: &GraphQLTypeAnnotation<SelectableServerFieldId>,
+    schema_data: &ServerFieldData,
+) -> GraphQLTypeAnnotation<UnvalidatedTypeName> {
+    type_.clone().map(|type_id| match type_id {
+        SelectionType::Scalar(scalar_id) => schema_data.scalar(scalar_id).name.item.into(),
+        SelectionType::Object(object_id) => schema_data.object(object_id).name.into(),
+    })
+}
+
+fn enum_satisfies_type(
+    enum_literal_value: &EnumLiteralValue,
+    enum_type: &GraphQLNamedTypeAnnotation<SelectableServerFieldId>,
+    schema_data: &ServerFieldData,
+    location: Location,
+) -> ValidateSchemaResult<()> {
+    match enum_type.item {
+        SelectionType::Object(object_id) => {
+            let expected = GraphQLTypeAnnotation::Named(GraphQLNamedTypeAnnotation(
+                enum_type
+                    .clone()
+                    .map(|_| schema_data.object(object_id).name.into()),
+            ));
+
+            Err(WithLocation::new(
+                ValidateSchemaError::ExpectedTypeFoundEnum {
+                    expected,
+                    actual: *enum_literal_value,
+                },
+                location,
+            ))
+        }
+        SelectionType::Scalar(_scalar_id) => {
+            todo!("Validate enum literal. Parser doesn't support enum literals yet")
+        }
+    }
+}
+
+fn list_satisfies_type(
+    list: &[WithLocation<NonConstantValue>],
+    list_type: GraphQLListTypeAnnotation<SelectableServerFieldId>,
+    variable_definitions: &[WithSpan<ValidatedVariableDefinition>],
+    schema_data: &ServerFieldData,
+) -> ValidateSchemaResult<()> {
+    list.iter().try_for_each(|element| {
+        value_satisfies_type(element, &list_type.0, variable_definitions, schema_data)
+    })
+}
+
+fn get_variable_type<'a>(
+    variable_name: &'a VariableName,
+    variable_definitions: &'a [WithSpan<ValidatedVariableDefinition>],
+    location: Location,
+) -> ValidateSchemaResult<&'a GraphQLTypeAnnotation<SelectableServerFieldId>> {
+    match variable_definitions
+        .iter()
+        .find(|definition| definition.item.name.item == *variable_name)
+    {
+        Some(variable) => Ok(&variable.item.type_),
+        None => Err(WithLocation::new(
+            ValidateSchemaError::UsedUndefinedVariable {
+                undefined_variable: *variable_name,
+            },
+            location,
+        )),
+    }
+}
+
+pub fn get_missing_arguments_and_validate_types(
+    schema_data: &ServerFieldData,
+    argument_definitions: &[&ValidatedVariableDefinition],
     arguments: &[WithLocation<SelectionFieldArgument>],
     include_optional_args: bool,
-) -> Vec<ValidatedVariableDefinition> {
+    variable_definitions: &[WithSpan<ValidatedVariableDefinition>],
+) -> ValidateSchemaResult<Vec<ValidatedVariableDefinition>> {
     argument_definitions
+        .iter()
         .filter_map(|definition| {
```

Comment:
nit: it's kind of hard to track what's going on here. Can we (throughout this file, and ideally as a separate PR) take a pass to rename variables to clearer?

e.g. `argument_definitions` -> `field_argument_definitions`, `argument_definitions_vec` -> `field_argument_definitions_vec`, `argument` -> `selection_supplied_arguments`, `user_supplied_argument` -> `selection_supplied_argument`